docs/adrs/1751-runner-job-hooks.md

```diff
@@ -0,0 +1,83 @@
+# ADR: Notification Hooks for Runners
+
+## Context
+
+This ADR details the design changes for supporting custom configurable hooks for on various runner events. This has been a long requested user feature [here](https://github.com/actions/runner/issues/1543), [here](https://github.com/actions/runner/issues/699) and [here](https://github.com/actions/runner/issues/1116) for users to have more information on runner observability, and for the ability to run cleanup and teardown jobs. 
+
+This feature is mainly intended for self hosted runner administrators.
+
+**What we hope to solve with this feature**
+1. A runner admininstrator is able to add custom scripts to cleanup their runner environment at the start or end of a job
+2. A runner admininstrator is able to add custom scripts to help setup their runner environment at the beginning of a job, for reasons like [caching](https://github.com/actions/runner/issues/1543#issuecomment-1050346279)
+3. A runner administrator is able to grab custom telemetry of jobs running on their self hosted runner
+
+**What we don't think this will solve**
+- Policy features that require certain steps run at the beginning or end of all jobs
+  - This would be better solved to in a central place in settings, rather then decentralized on each runner. 
+  - The Proposed `Notification Hooks for Runners` is limited to self hosted runners, we don't beileve Policy features should be
+- Reuse scenarios between jobs are covered by [composite actions](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) and [resuable workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows)
+- Security applications, security should be handled on the policy side on the server, not decentralized on each runner
+
+## Hooks
+- We will expose 2 variables that users can set to enable hooks  
+  - `ACTIONS_RUNNER_HOOK_JOB_STARTED`
+  - `ACTIONS_RUNNER_HOOK_JOB_COMPLETED`
+
+You can set these variables to the **absolute** path of a a `.sh` or `.ps1` file.
+
+We will execute `pwsh` (fallback to `powershell`) or `bash` (fallback to `sh`) as appropriate.
+- `.sh` files will execute with the args `-e {pathtofile}`
+- `.ps1` files will execute with the args `-command \". '{pathtofile}'\"`
+
+We will **not** set the [standard flags we typically set](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell) for `runs` commands. So, if you want to set `pipefail` on `bash` for example, you will need to do that in your script.
+
+### UI
+We want to ensure the experience for users invoking workflows is good, if hooks take too long, you may feel your job is delayed or broken. So, much like `Set Up Job`, we will generate two new steps automatically in your job, one for each configured hook: 
+- `Pre Job Hook`  
+- `Post Job Hook` 
```

Comment:
I'd like to tweak the verbiage here a little bit. I think that the runner admins who set these jobs up will understand what this means, but the end-user who is writing/executing a workflow on that runner may not.

We already have "Set up job" and "Complete job".  🤔 

I tend towards "Set up runner" and "Complete runner" to align? I don't _love_ "complete runner", because it's vague, but I like that it's matching.

Reply:
From above it seems like this feature is going to always be exclusive to self-hosted runners, right?

If that's the case, speaking as someone who is responsible for an org that uses self-hosted runners a bunch, it might help to have it be something like "Set up _self-hosted_ runner".

This is a minor point, but our users are not necessarily clear on where our infrastructure ends and where GitHub's begins... if they see weird stuff in the logs for a step about self-hosted runners they'll know who to blame quicker :) This isn't a big deal though, when they ask about something going wrong the people who can diagnose this look under "Set up job" to see who's runner it is, etc.

Reply:
> If that's the case, speaking as someone who is responsible for an org that uses self-hosted runners a bunch, it might help to have it be something like "Set up self-hosted runner".

That's definitely a scenario I've been thinking about a lot. The user who owns the self hosted runners may not always be the same user who is writing the workflows.

If I (as a workflow author) had my job fail because the runner admin injected a step that isn't working correctly, its not necessarily clear who I should contact to fix that. `Set up runner` sounds like it could be an issue with the runner application, I may not have context here.

To help, we are planning to add text that says `Running the 'Set up runner' step configured by the self-hosted runner administrator` as the first line of the step logs, to make it more clear  why this step was injected, and what users can do if it starts failing.